lib/rules/for-direction.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * @fileoverview enforce "for" loop update clause moving the counter in the right direction.(for-direction)
+ * @author Aladdin-ADD<hh_2013@foxmail.com>
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "enforce \"for\" loop update clause moving the counter in the right direction.",
+            category: "Possible Errors",
+            recommended: false
+        },
+        fixable: null,
+        schema: []
+    },
+
+    create(context) {
+
+        /**
+         * report an error.
+         * @param {ASTNode} node the node to report.
+         * @returns {void}
+         */
+        function report(node) {
+            context.report({
+                node,
+                message: "enforce \"for\" loop update clause moving the counter in the right direction."
+            });
+        }
+
+        /**
+         * check UpdateExpression add/sub the counter
+         * @param {ASTNode} update UpdateExpression to check
+         * @param {string} counter variable name to check
+         * @returns {int} if add return 1, if sub return -1, if nochange, return 0
+         */
+        function getUpdateDirection(update, counter) {
+            if (update.argument.name === counter) {
+                if (update.operator === "++") {
+                    return 1;
+                }
+                if (update.operator === "--") {
+                    return -1;
+                }
+            }
+            return 0;
+        }
+
+        /**
+         * check AssignmentExpression add/sub the counter
+         * @param {ASTNode} update AssignmentExpression to check
+         * @param {string} counter variable name to check
+         * @returns {int} if add return 1, if sub return -1, if nochange, return 0
+         */
+        function getAssignmentDirection(update, counter) {
+            if (update.left.name === counter && update.right.type === "Literal") {
+                if (update.operator === "+=") {
+                    return 1;
+                }
+                if (update.operator === "-=") {
+                    return -1;
+                }
+            }
+            return 0;
+        }
+        return {
+            ForStatement(node) {
+
+                if (node.test && node.test.type === "BinaryExpression" && node.update) {
+                    const counter = node.test.left.name;
```

Comment:
This code accesses `node.test.left.name`, but it never verifies that `node.left.type === "Identifier"`. So the rule will incorrectly report an error for something like this, because `node.left.name` is `undefined`:

```js
for (; foo.bar < 10; baz.qux--);
```

To fix this, I think it would be best to assert that `node.left.type === "Identifier"` in the `if` statement.

It would also be good to check the type of `update.left` and `update.argument` in the helper functions before accessing the `name` property.